include/nbl/builtin/glsl/format/constants.glsl

```diff
@@ -30,4 +30,10 @@
 #define nbl_glsl_RGB18E7S3_COMPONENT_BITOFFSETS ivec4(0,nbl_glsl_RGB18E7S3_MANTISSA_BITS,(2*nbl_glsl_RGB18E7S3_MANTISSA_BITS)&31,(3*nbl_glsl_RGB18E7S3_MANTISSA_BITS)&31)
 #define nbl_glsl_RGB18E7S3_G_COMPONENT_SPLIT (32-nbl_glsl_RGB18E7S3_MANTISSA_BITS)
 
+// From: https://www.khronos.org/opengl/wiki/Small_Float_Formats#R11F_G11F_B10F
+#define nbl_glsl_R11G11B10_MIN (6.10 * 1e-5)
+#define nbl_glsl_R11G11B10_MAX (6.50 * 1e4)
+#define nbl_glsl_R11G11B10_BIAS 15
+#define nbl_glsl_R11G11B10_EXPONENT_BITS 5
```

Comment:
these are awful names for constants, also they shouldn't be here/exist

you're using them for `encode_ufloat` (which encodes a single float) nor for r11g11b10

also don't trust some randos that wrote the GL wiki, read the spec
https://www.khronos.org/registry/DataFormat/specs/1.3/dataformat.1.3.inline.html#11bitfp

make a set of functions
```
uint nbl_glsl_ieee754_exponent_bias(in uint exponentBits)
{
    return (0x1u<<(exponentBits-1))-1;
}

float nbl_glsl_ieee754_true_min(in uint exponentBits, in uint mantissaBits)
{
    return exp2(1-int(nbl_glsl_ieee754_exponent_bias(exponentBits))-mantissaBits);
}

float nbl_glsl_ieee754_min(in uint exponentBits, in uint mantissaBits)
{
    const float e = exp2(1-int(nbl_glsl_ieee754_exponent_bias(exponentBits)));
    const uint m = 0x1u<<(23-mantissaBits);
    return uintBitsToFloat(floatBitsToUint(e)|m);
}

float nbl_glsl_ieee754_max(in uint exponentBits, in uint mantissaBits)
{
    const float e = exp2(1-int(nbl_glsl_ieee754_exponent_bias(exponentBits)));
    const uint m = 0x7fFFffu&(0x7fFFffu<<(23-mantissaBits));
    return uintBitsToFloat(floatBitsToUint(e)|m);
}
```
and use them throughout

Reply:
What is the difference between `nbl_glsl_ieee754_true_min` and `nbl_glsl_ieee754_min`?

Reply:
Actually I have no clue what any of this is doing.

I'm sure it does _not_ replace the current defined values of `nbl_glsl_R11G11B10_MIN` and `nbl_glsl_R11G11B10_MAX` (which I agree are horrible names), in the sense that:
```GLSL
nbl_glsl_ieee754_min(5, 6) != min possible value of 11bit float, currently approximated by nbl_glsl_R11G11B10_MIN
nbl_glsl_ieee754_max(5, 6) != max possible value of 11bit float, currently approximated by nbl_glsl_R11G11B10_MAX
```

Reply:
If we're indeed trying to do this, then I think this would be the correct way to do it:

```GLSL

// uint nbl_glsl_ieee754_exponent_bias(in uint exponentBits) remains the same

// float nbl_glsl_ieee754_true_min(in uint exponentBits, in uint mantissaBits) remains the same, if it
// corresponds to the E=0, M != 0 case in the spec you mentioned

float nbl_glsl_ieee754_min(in uint exponentBits, in uint mantissaBits)
{
    const uint biasSinglePrecision = nbl_glsl_ieee754_exponent_bias(8);
    const uint bias = nbl_glsl_ieee754_exponent_bias(exponentBits);

    const uint biasedMinExp = 1;
    const uint e = (biasedMinExp - bias) + biasSinglePrecision;
    const uint m = 0x1u << (23 - mantissaBits);

    return uintBitsToFloat((e << 23) | m);
}

float nbl_glsl_ieee754_max(in uint exponentBits, in uint mantissaBits)
{
    const uint biasSinglePrecision = nbl_glsl_ieee754_exponent_bias(8);
    const uint bias = nbl_glsl_ieee754_exponent_bias(exponentBits);

    const uint biasedMaxExp = (((1 << exponentBits) - 1) - 1); // `(1 << exponentBits) - 1` is reserved for Inf/NaN.
    const uint e = (biasedMaxExp - bias)+biasSinglePrecision;
    const uint m = 0x7fFFffu & (0x7fFFffu << (23 - mantissaBits));

    return uintBitsToFloat((e << 23) | m);
}
```

Reply:
> What is the difference between `nbl_glsl_ieee754_true_min` and `nbl_glsl_ieee754_min`?

normalized vs denormalized minimum value, its important to distinguish as denormalized:
- may flush to 0 and you end up with 0 instead of a positive value
- when an impl does not flush denorm to 0, floating point operations are much slower (go through a semi-software path)

Reply:
> If we're indeed trying to do this, then I think this would be the correct way to do it:
> 
> ```glsl
> // uint nbl_glsl_ieee754_exponent_bias(in uint exponentBits) remains the same
> 
> // float nbl_glsl_ieee754_true_min(in uint exponentBits, in uint mantissaBits) remains the same, if it
> // corresponds to the E=0, M != 0 case in the spec you mentioned
> 
> float nbl_glsl_ieee754_min(in uint exponentBits, in uint mantissaBits)
> {
>     const uint biasSinglePrecision = nbl_glsl_ieee754_exponent_bias(8);
>     const uint bias = nbl_glsl_ieee754_exponent_bias(exponentBits);
> 
>     const uint biasedMinExp = 1;
>     const uint e = (biasedMinExp - bias) + biasSinglePrecision;
>     const uint m = 0x1u << (23 - mantissaBits);
> 
>     return uintBitsToFloat((e << 23) | m);
> }
> 
> float nbl_glsl_ieee754_max(in uint exponentBits, in uint mantissaBits)
> {
>     const uint biasSinglePrecision = nbl_glsl_ieee754_exponent_bias(8);
>     const uint bias = nbl_glsl_ieee754_exponent_bias(exponentBits);
> 
>     const uint biasedMaxExp = (((1 << exponentBits) - 1) - 1); // `(1 << exponentBits) - 1` is reserved for Inf/NaN.
>     const uint e = (biasedMaxExp - bias)+biasSinglePrecision;
>     const uint m = 0x7fFFffu & (0x7fFFffu << (23 - mantissaBits));
> 
>     return uintBitsToFloat((e << 23) | m);
> }
> ```

my way should have been identical, because `exp2` with an integer argument should just give you a float with a "cleared" mantissa, so only the mantissa needs to be OR-ed in

If you want to do it your way then please define a 
```cpp
float nbl_glsl_exp2(in int x);
```
and use that to get your `e`

Reply:
> my way should have been identical, because exp2 with an integer argument should just give you a float with a "cleared" mantissa, so only the mantissa needs to be OR-ed in

Indeed and I think its better/more succinct.
However, calculation of `e` in `nbl_glsl_ieee754_max` suffers from copypasta (assuming), it should've been:
```GLSL
const uint biasedMaxExp = (((1 << exponentBits) - 1) - 1); // `(1 << exponentBits) - 1` is reserved for Inf/NaN.
const float e = exp2(biasedMaxExp - int(nbl_glsl_ieee754_exponent_bias(exponentBits)));
```

Reply:
> when an impl does not flush denorm to 0, floating point operations are much slower (go through a semi-software path)

Does that mean we use `nbl_glsl_ieee754_true_min` [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/5374db1c4b601e8a11d057ea359cb45440f0fa42/include/nbl/builtin/glsl/format/encode.glsl#L93)?
```GLSL
if (_f32 < uintBitsToFloat(maxSinglePrecisionVal))
{
    if (_f32 < nbl_glsl_ieee754_true_min(exponentBits, mantissaBits))
        return 0;
...
```
or `nbl_glsl_ieee754_true_min` is just a helper function for later and is not required anywhere right now?

Reply:
> > when an impl does not flush denorm to 0, floating point operations are much slower (go through a semi-software path)
> 
> Does that mean we use `nbl_glsl_ieee754_true_min` [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/5374db1c4b601e8a11d057ea359cb45440f0fa42/include/nbl/builtin/glsl/format/encode.glsl#L93)?
> 
> ```glsl
> if (_f32 < uintBitsToFloat(maxSinglePrecisionVal))
> {
>     if (_f32 < nbl_glsl_ieee754_true_min(exponentBits, mantissaBits))
>         return 0;
> ...
> ```
> 
> or `nbl_glsl_ieee754_true_min` is just a helper function for later and is not required anywhere right now?

you can use `true_min` if you can check that the rest of the code won't blow up if `_f32` is 0 and if didn't return 0 because the constant on the right side of the `<` flushed to 0.

Reply:
Doing this gives worse RMSE, so won't do.

Reply:
> Doing this gives worse RMSE, so won't do.

you mean using `true_min` instead of `min` ?